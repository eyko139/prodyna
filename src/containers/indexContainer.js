import React, { useState, useEffect } from "react";
import Input from "../components/input";
import { BrowserRouter,
  Switch,
  Route,
  Link } from "react-router-dom";

function IndexContainer() {

  async function postData(url = "", data = {}) {
    const response = await fetch(url, {
      method: "POST",
      headers: {
        "Content-Type": "application/json"
      },
      body: JSON.stringify(data)
    })
    .then(response => {
      if (response.ok) {
        return response.json()
      }
      throw new Error("Request failed!")
    }, networkError => console.log(networkError.message))
    return response;
    //Insert a render response function?
  }

  const [result, setResult] = useState("");

  const handleSubmit = (event) => {
    event.preventDefault()
    postData("/api/input", {
      text,
      letter,
      caseSensitivity,
    })
    .then(data => {
      console.log(data);
      setResult(data.occurances)
      window.location.href = data.url;
    });
    
  }
  const [caseSensitivity, setCaseSensitivity] = useState(false);
  const handleCase = ({ target }) => {
        caseSensitivity ? setCaseSensitivity(false) : setCaseSensitivity(true);
    }

  const [text, setText] = useState("")
  const handleTextChange = ({ target }) => {
    setText(target.value);
  }

  const [letter, setLetter] = useState("")
  const handleLetterChange = ({ target }) => {
    setLetter(target.value)
  }

  // useEffect(() => {
  //     console.log(caseSensitivity);
  // }, [caseSensitivity])
  // useEffect(() => {
  //     console.log(text)
  // }, [text, letter])
  
  return (
      <Input
        handleTextChange={handleTextChange}
        handleLetterChange={handleLetterChange}
        handleCaseChange={handleCase}
        handleSubmit={handleSubmit}
        caseSensitivity={caseSensitivity}
        text={text}
        letter={letter}
        caseSensitivity={caseSensitivity}
      />
  )
}

  export default IndexContainer;