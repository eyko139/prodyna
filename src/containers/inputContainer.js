import React, { useState, useEffect, useRef } from "react";
import Input from "../components/inputForm";
import validate from "./validateInput";
import useFirstRender from "./validationHook";

function IndexContainer() {

  //User Input
  const [values, setValues] = useState({
    text: "",
    letter: "",
    caseSensitivity: false
  })

  //Form validation errors
  const [validationErrors, setValidationErrors] = useState({});

  //custom hook returns true for first render, false for every subsequent render
  let firstRender = useFirstRender()
  useEffect(() => {
    if (!firstRender || validationErrors) {
      setValidationErrors(validate(values))
      console.log(validationErrors)
    }
  },[firstRender, values])

  //Sets the corresponding values for all input forms 
  const handleChange = ({ target }) => {
    const { name, value } = target
    if (target.name === "text" || target.name === "letter") {
      setValues({
        ...values,
        [name]: value
      })
    }
    else {
      target.checked ? setValues({...values, caseSensitivity: true}) : setValues({...values, caseSensitivity: false})
    }
  };

  //API put request, called on form submission
  async function postData(url = "", data = {}) {
    try {
      const response = await fetch(url, {
        method: "PUT",
        headers: {
          "Content-Type": "application/json"
        },
        body: JSON.stringify(data)
      })
      if (response.ok) {
          const jsonResponse = await response.json()
          return jsonResponse;
        }
      throw new Error("Request failed!")
    }
    catch(netWorkError) {
      console.log(netWorkError)
    }
  }

  //Form submission:
  //Checks for errors and runs validation check in case there was no input
  const handleSubmit = (event) => {
    event.preventDefault()

    if (Object.keys(validationErrors).length !== 0 || values.letter == "") {
      setValidationErrors(validate(values))
      return
    }
    postData("/api", {
      apitext: values.text,
      apiletter: values.letter,
      case_sensitive_search: values.caseSensitivity,
      key: Date.now(),
    })
    .then(data => {
      window.location.href = data.url;
    });
    
  }

  
  return (
    <div data-testid="inputForm">
      <Input
        handleSubmit={handleSubmit}
        caseSensitivity={values.caseSensitivity}
        text={values.text}
        letter={values.letter}
        handleChange={handleChange}
        errors={validationErrors}
      />
    </div>
  )
}

  export default IndexContainer;