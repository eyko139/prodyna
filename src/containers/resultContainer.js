import React, { useState, useEffect } from "react"

import Result from "../components/result.js"
import SingleResult from "../components/singleResult.js"
function ResultContainer() {
    const [results, setResults] = useState([])
    const [loading, setLoading] = useState(false)

    useEffect(() => {
        const fetchResults = async() => {
            setLoading(true)
            const response = await fetch("/hello", {
              method: "GET"
            })
            .then(response => {
               if (response.ok) {
                    return response.json()}
               })
            .then(response => {
                console.log(response);
                return response
            })
            setResults(response)
        }
        fetchResults();
    }, [])
    console.log(results)

    const deleteResultApi = async(keyToRemove) => {
        const response = await fetch("/hello", {
            method: "DELETE",
            headers: {
                "Content-Type": "application/json"
            },
            body: JSON.stringify({id: keyToRemove})
        })
    }

    const deleteResult = (index, keyToRemove) => {
        setResults(results.filter(result => result.id != keyToRemove))
        deleteResultApi(index)
    }
    return ( 
        <div>
            {/* <Result
              results={results}
              deleteResult={deleteResult}
            /> */}

            { results.map((result, index) => (
                <SingleResult 
                    letter={result.letter}
                    text={result.text}
                    id={result.id}
                    letterOccurances={result.letter_occurances}
                    is_case_sensitive={result.is_case_sensitive}
                    index={index}
                    deleteResult={deleteResult}
                />
            ))}
        </div>
    )
}
export default ResultContainer;