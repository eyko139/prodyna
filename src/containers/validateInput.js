import React from "react"

export default function validateInfo(values) {
    let errors = {};
    const letterInput = document.getElementById("letterInput")
    const textInput = document.getElementById("textInput")
  

    //Error for empty or non-alpha letter input
    if (!/^[a-z]/i.test(values.letter)) {
        errors.letter= "Only letters";
        errors.field = "letter"
        letterInput.setCustomValidity("Only letters")
    }  
    if (values.letter === "") {
        errors.letter = "Field cannot be empty";
        errors.field = "letter"
        letterInput.setCustomValidity("Field cannot be empty")
    }
    if (/\s/.test(values.letter)) {
      errors.letter = "No spaces";
      errors.field = "letter";
      letterInput.setCustomValidity("Cant be only space!")
    }

    if (values.text === "" || !/[a-z]/i.test(values.text)) {
      errors.text = "Field cannot be empty";
      errors.field = errors.field + " text";
      textInput.setCustomValidity("Field cannot be empty")
    }
    
    if (!errors.letter) {
      letterInput.setCustomValidity("")
    }
    if (!errors.text) {
      textInput.setCustomValidity("")
    }


    return errors
}