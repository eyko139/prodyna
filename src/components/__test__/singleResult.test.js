import React from "react";
import SingleResult from "../singleResult";
import { render, fireEvent, cleanup } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect"
import { deleteResult } from "../../containers/resultContainer";
import renderer from "react-test-renderer";
import {screen} from '@testing-library/dom';

afterEach(cleanup);
//Testing result rendering
it("Result renders correctly", () => {
    const { getByTitle, getByText, getByRole, getAllByRole } = render(<SingleResult 
                                letter="a"
                                text="This is a test text"
                                id="1"
                                letterOccurances="1"
                                is_case_sensitive="true"
                                index="0"
                                deleteResult={deleteResult}
                                    />);
    const container = getByRole("container");
    const header = getByRole("heading", {level: 2})
    const buttons = getAllByRole("button")
    expect(container).toBeTruthy();
    expect(header.innerHTML).toBe("The letter \"a\" occured <p id=\"resultCounter\">1</p>time (case sensitive)");
    expect(buttons).toBeTruthy();
})

//Testing highlight function
describe("clickHighlight", () => {
    it("onClick", () => {
        const renderFn = jest.fn().mockReturnValue(null);
        const wrapper = render(<SingleResult
                                letter="a"
                                text="Thisisatesttext"
                                id={Date.now()}
                                letterOccurances="1"
                                is_case_sensitive="true"
                                index="0"
                                deleteResult = {deleteResult}
                                />)

        // const { getByText, getAllByTitle, getByTitle} = render(<SingleResult 
                                // letter="a"
                                // text="This is a test text"
                                // id="1"
                                // letterOccurances="1"
                                // is_case_sensitive="true"
                                // index="0"
                                // deleteResult = {deleteResult}
        //                             />);
        const highlightButton = wrapper.getByTitle("highlightButton");
        expect(highlightButton).toBeTruthy();
        const searchedTExt = wrapper.getByTestId("searched");
        let highlightedLetters;
        setTimeout(() => {
            expect(searchedTExt.textContent).toBe("Thisisatesttext");
            highlightedLetters = screen.getAllByTitle("highlight");
            console.log(highlightedLetters)
        }, 150)
        fireEvent.click(highlightButton);
        expect(highlightedLetters.style.color).toBe("red");
    });
});

// test("Test 1", () => {
//     const component = renderer.create(
//       <SingleResult 
//                                 letter="a"
//                                 text="This is a test text"
//                                 id="1"
//                                 letterOccurances="1"
//                                 is_case_sensitive="true"
//                                 index="0"
//                                 deleteResult={deleteResult}
//                                 />
//     );
  
//     let tree = component.toJSON();
//     expect(tree).toMatchSnapshot();
//   });