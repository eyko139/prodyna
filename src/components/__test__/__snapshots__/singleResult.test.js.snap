// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test 1 1`] = `
<div
  className="resultContainer"
  role="container"
>
  <div
    className="resultCard"
    id="1"
  >
    <h2>
      The letter "
      a
      " occured 
      <p
        id="resultCounter"
      >
        1
      </p>
      time 
      (
      case sensitive
      )
    </h2>
    <div
      className="recap"
    >
      <div
        className="searchedText"
      >
        <p>
          Searched text:
        </p>
        <div
          className="textParagraph"
          id="p1"
          role="searchedText"
        >
          This is a test text
        </div>
      </div>
    </div>
    <div
      className="buttons"
    >
      <button
        id="delete"
        onClick={[Function]}
        title="highlightButton"
      >
        Delete
      </button>
      <button
        id="highlightButton"
        onClick={[Function]}
        title="deleteButton"
      >
        Hightlight Letter
      </button>
    </div>
  </div>
</div>
`;
